### PR TITLE
[PRFC] Release Process & Versioning Policy

### PR DESCRIPTION
**Pull request by Rugvip**
Co-authored-by: Johan Haals <johan.haals@gmail.com>

**Status:** Open for comments

## Need

Backstage is approaching its first distinct release, and as part of this we are looking to update the release schedule and clarify the versioning policy. We also see the need for the introduction of multiple release lines as we plan to slow down the release schedule, but still want to make features available early for experimentation.

Our current Stability Index operates at a per-package level or groups of APIs, but we feel that this is both too granular and too coarse. We instead look to create an overarching policy that applies to all packages in a more straight-forward way, but to make it possible for individual exports to be marked as unstable.

A piece that is missing in our current documentation is any kind of version skew policy. That is what are the recommended relations and maximum differences in versions between components running in the same environment?

## Proposal

This versioning policy proposal introduces two release lines, `main` and `next`. The current weekly release is replaced by the `next` release line that publishes `x.y.z-next.n` package versions of all packages with pending changes. Once every month the main-line release goes out, where each packages is promoted to a stable release, for example from `1.2.3-next.4` to `1.2.3`. The main-line releases replace the `next` release, meaning there will be no `next` release that week.

We are also introducing a new umbrella version of Backstage called a "Backstage release", which is a version assigned to each main-line release. The version format is `<major>.<minor>.<patch>`, but it does not follow semver. In this PR we suggest a new versioning policy as well as a version skew policy, both based on this release plan.

The primary purpose of the releases are to serve as a communication tool and provide context. Rather than referring to individual package versions you can now talk about a release version instead, for example "Backstage 1.2". We do however plan to introduce the release concept into some of our tooling such as `backstage-cli versions:bump`, which will be updated to be able to track a release line. Users that want to run the most recent version of Backstage might for example be able to use the `--release-line next` flag, or you might want to follow the main-line releases using `--release-line main`.

## Alternatives

This proposal is based on the desire to keep development in the repo simple, with a single branch used for development. We include emergency patch releases in the plan, but they are done on branches, and rarely. We've also experimented with other options than putting the entire repo in pre-release mode, but found that none of those options worked particularly well. It essentially ends up mimicking a multi-repo setup in a single repo, with all of the drawbacks and none of the benefits.

## Risks

There is a risk that only releasing main-line releases once a month, or even more seldom in the future, will have an impact on the development of plugins in the main repo. One of the reasons we introduce the `next` release line is to counteract this, allowing plugins to be released with the same cadence as today. This also serves to make sure that the `next` release line is used and that bugs can be found before they make it into a main-line release. A downside of this is that some plugins may end up being primarily consumed through pre-releases, which do not provide the same stability guarantees as our current releases. In order to decouple plugins from the main-repo release process, we may set up separate repositories in the `backstage` GitHub organization in the future, but that is not included as part of the initial proposal. 

Another risk that we've identified is sticking to a single release track, and what impact that will have both on contributions as well as stability. It may be that we will want to add additional policies in the future that steer when and what changes we make leading up to a main-line release, in order to avoid introducing bugs just before one goes out.

### CODE REVIEW

**Comment by Xantier** [comment]
I think some more context around this proposal might be useful here. Few questions come to mind around this: 

What does the release version signify? Does that point to the "whole Backstage" as an individual deployable blob (a scaffolded app), or does the number leak into individual larger features (catalog in version 1.2, scaffolder in 1.1 etc). How do contributors tie the umbrella version to individual packages and their semver versions?

Does it make sense to use 'semver-like' release versioning here, since the significance of that semver value is not followed. What would it mean to bump the first number from 1 to 2? Does technical functionality (+possible large visual changes) like bumping MUI to a next version achieve that threshold? What does it mean for minor numbers?

Would it make sense to use time based versioning system instead since the cadence of releases will be monthly? Something like `2022.3.3`, where major version is March 2022, and this revision has received 3 critical hotfixes. With this those faster cadence weeklies can be identified by something like `2022.4-next.3`.

**Comment by Rugvip** [comment]
> What does the release version signify? Does that point to the "whole Backstage" as an individual deployable blob (a scaffolded app), or does the number leak into individual larger features (catalog in version 1.2, scaffolder in 1.1 etc). How do contributors tie the umbrella version to individual packages and their semver versions?

Tbh communication-wise that's not entirely figured out yet. It may be that we end up talking about "catalog 1.2" meaning the catalog that is part of the `1.2.0` Backstage release. It could also be that "catalog 1.2" is decoupled and released in `1.4.0` or what have you. I'm a bit worried that the second options would create too much clutter, as we then have 3 different types of versions to keep track of. Either way, not figured out yet and happy to discuss ideas :grin:

Something that we may end up doing is treating `>=1.0` packages as a bit more special in each release. To the point where we don't even consider `0.x` packages to be part of the release, even though they ofc technically will be published together. This would again mostly apply to communication around releases rather than having any technical impact.

Regarding tying umbrella version to individual packages, we plan to ship a package that contains that mapping. With that we have the data that's needed and can create any additional tools and utilities on top of that.

> Does it make sense to use 'semver-like' release versioning here, since the significance of that semver value is not followed. What would it mean to bump the first number from 1 to 2? Does technical functionality (+possible large visual changes) like bumping MUI to a next version achieve that threshold? What does it mean for minor numbers? Does it make sense to have these 2 numbers as separate? Major = a bunch of changes, minor = a (smaller) bunch of changes.

This is all inspired by the [Kubernetes releases](https://kubernetes.io/releases/), who still is on `1.x` after plenty of big releases :grin: We cover this a bit in the policy but what the major version really means is TBD and it's reserved for future use at this point. Something like bumping MUI is unlikely to move that version.

> Would it make sense to use time based versioning system instead since the cadence of releases will be monthly? Something like `2022.3.3`, where major version is March 2022, and this revision has received 3 critical hotfixes. With this those faster cadence weeklies can be identified by something like `2022.4-next.3`.

The initial release cadence is probably monthly, but I think it's possible that we want to slow it down even more in the future. Whether we do that is gonna end up depending on a bunch of factors, but the thought right now is to stick to a simple incrementing number that isn't tied directly to any dates. I think there's a a bit of a risk of confusion if we start by releasing `22.1`, `22.2`, `22.3`, `22.4`, but then for example jump to `22.6`, `22.8`, `22.10`.

**Comment by iamEAP** [comment]
> _...version skew policy. That is what are the recommended relations and maximum differences in versions between components running in the same environment?_

Wondering about this in particular.  Is the intention for individual plugins to be able to declare, "I am compatible with Backstage ^1.2" via some means separate from package versioning / dependencies?  Or am I misunderstanding?  Some complications that come to mind:

- If the version numbers _aren't_ semantic in any way, how would a plugin indicate "I don't work with versions prior to 1.2, I _do_ work with versions 1.2 and up, but I do not work with version 2.0 or up"
- How would a plugin developer go about verifying that their plugin works against different versions of Backstage (automated and manual)?
- Would this imply we would be loosening the cross-package version bumps we do now (where a patch change in `plugin-x` results in a corresponding `package.json` version increment in `plugin-y` that depends on `plugin-x`?

**Comment by Rugvip** [comment]
@iamEAP yeah let's keep the discussion up for how exactly we want to name the versions. I like the current proposal because it gives us a nice looking version, but also a tool for going "hey look at this, this version is special!" with a major bump.

For now the skew policy really only covers what's in the doc, so nothing specific between plugins or for plugins to declare. I think this release process and some of the tooling we put in place can help out though, so perhaps it's something that could evolve in the future.

**Comment by awanlin** [comment]
We've got a semi-automated process in place for upgrading Backstage. Right now it just runs on a schedule every Friday morning as we know a release would have been done on Thursday. All it does is create a branch, run the version bump commit the changes and then create a PR which triggers the related PR build which catches any issues.

Now for part of that we use the GitHub API to determine the release name as we use that for the branch name and have plans to also create a git tag with it. Will we be able to detect the new  `main` and `next` releases lines from there? I have a feeling we'd be looking at the release name itself? The API results do have a `prerelease` flags. This is the endpoint we have been using: https://api.github.com/repos/backstage/backstage/releases/latest

File: docs/overview/versioning-policy.md [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+id: versioning-policy
+title: Release & Versioning Policy
+description:
+---
+
+The Backstage project is comprised of a set of software components that together
+form the Backstage platform. These components are both plugins as well as core
+platform libraries and tools. Each component is distributed as a collection of
+[packages](<https://en.wikipedia.org/wiki/Npm_(software)>), which in the end is
+what you end up consuming as an adopter of Backstage
```
[the remark this answers is not in the document]
Comment (Rugvip) [reply docs/overview/versioning-policy.md]: yep! 👍

**Comment by Rugvip** [comment]
@awanlin yep, we'd mark the GitHub releases that go out on the next line as prereleases.

We also plan to make some form of API available that you can use to look up versions in tooling. We want something ourselves that we can use to switch to specific releases using `backstage-cli versions:bump`.

**Review by freben — APPROVED** [review]

**Comment by Xantier** [comment]
Returning to this one while it's been brewing in production for few weeks now. If we take a look at the versioning policy of individual packages [within the latest pre-release](https://github.com/backstage/backstage/releases/tag/v1.1.0-next.3), we can see that there have been some introductions of breaking changes within minor version bumps to individual packages. 

I think this might accidentally start breaking a lot of applications in the wild. A lot of Backstage packages have now graduated to the above the 1.0.0 threshold world. That combined with the unfortunate tendency to use caret `^` ranges JS in the ecosystem as well as introduced breaking changes in minor package versions will pull in unknown, unexpected, breaking changes.

Just a reminder, I know this is well known, but pre 1.0.0. caret will not pull in minor changes so adding breakages on those will be fine from users perspective. After 1.0.0 caret range will pull in those minor changes and we shouldn't be having breakages between those versions, major should be bumped instead.

**Comment by jhaals** [comment]
Looks like some techdocs packages contain breaking changes in the latest pre release. @backstage/techdocs-core should probably verify the breaking change and see if this warrant a 2.0 or if other changes can be made prior to the release.

tl;dr of the policy:
Packages on version 0.x is still allowed to introduce breaking changes in minor versions.
Packages above 1.x with breaking changes should be major bumped.

@Xantier I hope I didn't miss anything your message.
